sql: panic: runtime error: index out of range

```
CREATE TABLE t (col2 INET);

ALTER TABLE t
    ADD COLUMN
        col_44353
            DECIMAL
            NULL
            CHECK (
                CASE
                WHEN (false AND false)
                THEN (
                    CASE
                    WHEN true
                    THEN 'defe:e132:a582:a462:953d:1da1:9a50:3cda/97':::INET
                    ELSE col2
                    END::INET
                    << (col2::INET | col2::INET)::INET
                )
                ELSE false
                END
            )
            CHECK (true)
            CHECK (false)
            CHECK (false)
            CHECK (false);
```

```
panic: runtime error: index out of range

goroutine 597 [running]:
github.com/cockroachdb/cockroach/pkg/sql/opt/memo.(*FiltersExpr).Child(0xc0019cada8, 0x0, 0x0, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/memo/expr.og.go:6231 +0x4f
github.com/cockroachdb/cockroach/pkg/sql/opt/idxconstraint.(*indexConstraintCtx).makeSpansForAnd(0xc0019cad48, 0x0, 0x39940e0, 0xc0019cada8, 0xc0019cadc0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/idxconstraint/index_constraints.go:690 +0x55
github.com/cockroachdb/cockroach/pkg/sql/opt/idxconstraint.(*indexConstraintCtx).makeSpansForExpr(0xc0019cad48, 0x0, 0x39940e0, 0xc0019cada8, 0xc0019cadc0, 0x1)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/idxconstraint/index_constraints.go:611 +0xd4c
github.com/cockroachdb/cockroach/pkg/sql/opt/idxconstraint.(*Instance).Init(0xc0019cad48, 0xc0019cc7e0, 0x0, 0x1, 0xc0019e4888, 0x1, 0x1, 0x4, 0x0, 0x1add600, ...)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/idxconstraint/index_constraints.go:1091 +0x273
github.com/cockroachdb/cockroach/pkg/sql.(*indexInfo).makeIndexConstraints(0xc0019cad20, 0xc0019f0900, 0xc0019cc7e0, 0x0, 0x1, 0xc000a5b510, 0x1, 0x6c636f)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt_index_selection.go:512 +0x4be
github.com/cockroachdb/cockroach/pkg/sql.(*planner).selectIndex(0xc000a5b450, 0x3986860, 0xc001a900c0, 0xc001830d00, 0x0, 0x7f8c153a5000, 0x2, 0xc00042c8c0, 0x1, 0x1)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt_index_selection.go:153 +0x1646
github.com/cockroachdb/cockroach/pkg/sql.expandScanNode(0x3986860, 0xc001a900c0, 0xc000a5b450, 0x1, 0x0, 0x0, 0x0, 0x0, 0xc001830d00, 0xc00042c8c0, ...)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/expand_plan.go:528 +0x101
github.com/cockroachdb/cockroach/pkg/sql.doExpandPlan(0x3986860, 0xc001a900c0, 0xc000a5b450, 0x1, 0x0, 0x0, 0x0, 0x0, 0x39885a0, 0xc001830d00, ...)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/expand_plan.go:317 +0x294b
github.com/cockroachdb/cockroach/pkg/sql.expandRenderNode(0x3986860, 0xc001a900c0, 0xc000a5b450, 0x1, 0x0, 0x0, 0x0, 0x0, 0xc0019c8900, 0xc00147bc90, ...)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/expand_plan.go:541 +0xfd
github.com/cockroachdb/cockroach/pkg/sql.doExpandPlan(0x3986860, 0xc001a900c0, 0xc000a5b450, 0x1, 0x0, 0x0, 0x0, 0x0, 0x39884e0, 0xc0019c8900, ...)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/expand_plan.go:320 +0x24f7
github.com/cockroachdb/cockroach/pkg/sql.doExpandPlan(0x3986860, 0xc001a900c0, 0xc000a5b450, 0x1, 0x0, 0x0, 0x0, 0x0, 0x39881e0, 0xc0019f8aa0, ...)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/expand_plan.go:238 +0x3176
github.com/cockroachdb/cockroach/pkg/sql.(*planner).expandPlan(0xc000a5b450, 0x3986860, 0xc001a900c0, 0x39881e0, 0xc0019f8aa0, 0x39881e0, 0xc0019f8aa0, 0x0, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/expand_plan.go:35 +0xc4
github.com/cockroachdb/cockroach/pkg/sql.(*planner).optimizePlan(0xc000a5b450, 0x3986860, 0xc001a900c0, 0x39881e0, 0xc0019f8aa0, 0xc0019e484a, 0x3, 0x3, 0x39881e0, 0xc0019f8aa0, ...)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/optimize.go:47 +0xeb
github.com/cockroachdb/cockroach/pkg/sql.(*planner).makePlan(0xc000a5b450, 0x3986860, 0xc001a900c0, 0x3220e7a, 0x12)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:359 +0x373
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).makeExecPlan(0xc000a5b100, 0x3986860, 0xc001a900c0, 0xc000a5b450, 0x6, 0x6c636f)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1012 +0x302
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc000a5b100, 0x3986860, 0xc001a900c0, 0xc000a5b450, 0x39aabe0, 0xc001a20060, 0x0, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:874 +0x148
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc000a5b100, 0x3986860, 0xc001a900c0, 0x398c460, 0xc0019f2540, 0xc0004c5900, 0xf6, 0x0, 0x0, 0x0, ...)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:459 +0xe03
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc000a5b100, 0x3986860, 0xc001a900c0, 0x398c460, 0xc0019f2540, 0xc0004c5900, 0xf6, 0x0, 0x0, 0x0, ...)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:101 +0x610
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc000a5b100, 0x3986860, 0xc0019c1ce0, 0xc00199cf00, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1189 +0x21c8
github.com/cockroachdb/cockroach/pkg/sql.(*internalExecutorImpl).initConnEx.func1(0xc000a5b100, 0x3986860, 0xc0019c1ce0, 0xc001985ec0, 0xc0019cf200, 0xc0019cc120, 0xc0019e4770)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:206 +0x64
created by github.com/cockroachdb/cockroach/pkg/sql.(*internalExecutorImpl).initConnEx
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:205 +0x4b3
```

--- a/pkg/sql/logictest/testdata/logic_test/check_constraints
+++ b/pkg/sql/logictest/testdata/logic_test/check_constraints
@@ -310,3 +310,19 @@ UPDATE t9 SET a = 7 WHERE a = 4
 
 statement error pgcode 23514 failed to satisfy CHECK constraint \(a > b\)
 UPDATE t9 SET a = 2 WHERE a = 5
+
+# Regression test for #36293. Make sure we don't panic with a false check
+# constraint.
+statement ok
+CREATE TABLE t36293 (x bool)
+
+statement ok
+ALTER TABLE t36293
+  ADD COLUMN y INT
+  CHECK (
+    CASE
+    WHEN false
+    THEN x
+    ELSE false
+    END
+  )

--- a/pkg/sql/opt/idxconstraint/index_constraints.go
+++ b/pkg/sql/opt/idxconstraint/index_constraints.go
@@ -604,12 +604,17 @@ func (c *indexConstraintCtx) makeSpansForExpr(
 
 	switch t := e.(type) {
 	case *memo.FiltersExpr:
-		if len(*t) == 1 {
+		switch len(*t) {
+		case 0:
+			c.unconstrained(offset, out)
+			return true
+		case 1:
 			return c.makeSpansForExpr(offset, (*t)[0].Condition, out)
+		default:
+			// We don't have enough information to know if the spans are "tight".
+			c.makeSpansForAnd(offset, t, out)
+			return false
 		}
-		// We don't have enough information to know if the spans are "tight".
-		c.makeSpansForAnd(offset, t, out)
-		return false
 
 	case *memo.FiltersItem:
 		// Pass through the call.

--- a/pkg/sql/opt/idxconstraint/index_constraints_test.go
+++ b/pkg/sql/opt/idxconstraint/index_constraints_test.go
@@ -152,6 +152,9 @@ func TestIndexConstraints(t *testing.T) {
 				}
 				root := f.Memo().RootExpr().(opt.ScalarExpr)
 				filters := memo.FiltersExpr{{Condition: root}}
+				if _, ok := root.(*memo.TrueExpr); ok {
+					filters = memo.TrueFilter
+				}
 
 				var ic idxconstraint.Instance
 				ic.Init(filters, indexCols, notNullCols, invertedIndex, &evalCtx, &f)

--- a/pkg/sql/opt/idxconstraint/testdata/misc
+++ b/pkg/sql/opt/idxconstraint/testdata/misc
@@ -251,3 +251,8 @@ index-constraints vars=(string) index=(@1)
 ----
 [/e'a\x00' - ]
 Remaining filter: length(@1) = 2
+
+index-constraints vars=(bool) index=(@1)
+true
+----
+[ - ]

